src/lib/linux-api/src/signal.rs

```diff
@@ -1097,6 +1183,142 @@ pub fn kill(pid: i32, sig: Option<Signal>) -> Result<(), Errno> {
     kill_raw(pid, sig.map(i32::from).unwrap_or(0))
 }
 
+/// Calls the `rt_sigaction` syscall.
+///
+/// # Safety
+///
+/// * `new_action` must be safe to dereference.
+/// * `old_action` must be safe to write to. (uninitd is ok).
+/// * `new_action`'s handler must be safe to call as a signal handler.
+///   See `signal-safety(7)`.
+/// * Generally, `new_action` must have `SA_RESTORER` set and a suitable
+///   `restorer`, such as [`sigaction_restorer`]. (There might be some esoteric
+///   way to call this syscall without this property, but I'm not aware of one).
+pub unsafe fn rt_sigaction_raw(
+    signo: i32,
+    new_action: *const sigaction,
+    old_action: *mut sigaction,
+    sigsetsize: usize,
+) -> Result<(), Errno> {
+    unsafe {
+        syscall!(
+            linux_syscall::SYS_rt_sigaction,
+            signo,
+            new_action,
+            old_action,
+            sigsetsize
+        )
+    }
+    .check()
+    .map_err(Errno::from)
+}
+
+/// Calls the `rt_sigaction` syscall.
+///
+/// # Safety
+///
+/// * `new_action`'s handler must be safe to call as a signal handler.
+///   See `signal-safety(7)`.
+/// * Generally, `new_action` must have `SA_RESTORER` set and a suitable
+///   `restorer`, such as [`sigaction_restorer`]. (There might be some esoteric
+///   way to call this syscall without this property, but I'm not aware of one).
+pub unsafe fn rt_sigaction(
+    signal: Signal,
+    new_action: &sigaction,
+    old_action: Option<&mut sigaction>,
+) -> Result<(), Errno> {
+    unsafe {
+        rt_sigaction_raw(
+            signal.as_i32(),
+            new_action,
+            old_action
+                .map(|o| o as *mut _)
+                .unwrap_or(core::ptr::null_mut()),
+            core::mem::size_of::<sigset_t>(),
+        )
+    }
+}
+
+#[cfg(test)]
+mod rt_sigaction_tests {
+    use core::sync::atomic::AtomicI32;
+
+    use shadow_pod::zeroed;
+
+    use super::*;
+
+    #[cfg(not(miri))]
+    #[test]
+    fn test_rt_sigaction() {
+        // Test signal handler
+        static CALL_COUNTER: AtomicI32 = AtomicI32::new(0);
+        extern "C" fn handler(_signo: i32) {
+            CALL_COUNTER.fetch_add(1, core::sync::atomic::Ordering::Relaxed);
+        }
+
+        // Test sigaction
+        let action = sigaction::new_with_default_restorer(
+            SignalHandler::Handler(handler),
+            SigActionFlags::empty(),
+            sigset_t::EMPTY,
+        );
+
+        // Signal that we'll be using.
+        let signal = Signal::SIGUSR2;
+
+        // Install our handler.
+        let mut old_action = zeroed();
+        unsafe { rt_sigaction(signal, &action, Some(&mut old_action)) }.unwrap();
```

Comment:
Maybe add a comment for this test that this handler is set globally across all tests, so if another test sets its own signal handler in another thread for this signal, there would be a race condition.

Reply:
done